``FpGroup`` implementation in SymPy

gxyd opened the pull request:
Implementation of classes/methods for "Coset Enumeration".
Currently very crude.

Commit:
``FpGroup`` implementation in SymPy

jksuom commented on sympy/combinatorics/free_group.py:
```diff
@@ -127,6 +127,7 @@ class FreeGroup(DefaultPrinting):
     """
     is_associative = True
     is_group = True
+    is_fp_group = True
```
There is `is_FpGroup` above. This might be confusing.

gxyd replied:
This is how I think about it. `.is_FpGroup` would return `True` for an instance of `FpGroup`, while `.is_fp_group` would be for mathematically checking whether the group is finitely presented or not. Since a free group is a finitely presented group with empty relators list, hence `is_fp_group = True` is set.
This is similar to `.is_number` and `.is_Number`.

leosartaj replied:
> This is how I think about it. .is_FpGroup would return True for an instance of FpGroup

I am -1 to this kind of functionality. I know this is abundantly done in SymPy codebase, but it is just not that useful. I think ``isinstance`` method covers all of the use cases. Plus, ``isinstance`` is always more clearer to me.

jksuom replied:
I was also going to suggest using `isinstance`.

gxyd replied:
I agree, so will remove "this" assignment (i.e `.is_fp_group = True`).

jksuom replied:
I'm not sure which assignment would be more useful. It seems that testing, e.g., `is_FpGroup` could take a little less time than `isinstance` but at least now I cannot think of a situation where such tests would be needed so frequently that the difference would become meaningful. In addition, there is the problem that the attribute `is_FpGroup` would probably have to be added also elsewhere. Otherwise it could not be used without first testing its existence.

leosartaj replied:
To me ``is_fp_group`` seems Ok. I am not a fan of ``is_FpGroup`` instead of using ``isinstance`` though.

Commit:
cyclic reduction methods added

Commit:
add methods "define" and "scan" from Derek Holt

Commit:
Add the "Coicidence" procedure

Commit:
Add "scan_and_fill" method

Commit:
Add methods for Felsch strategy

Commit:
Add more docs

gxyd commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -0,0 +1,336 @@
+            return
+    b = alpha
+    j = r
+    while j >= i and C[b][A.index(word.subword(j-1, j).inverse())] != -1:
```
I am looking at the Pg. 155 of Derek Holt, pseudo-code for scan line number 9-10 there is "b​^{x^​−​​1}"​​ is defined"
According to the Derek Holt. pseudo-code is
```
while j >= i and b^{x^-1} is defined
    do; b:=b^{x^-1}; j:=j–1;
```

My point is in book it uses `b^{x^-1}`, which I initially fortunately/unfortunately coded as it is my commit right now. Also what we have commited here seems to work, but the one in book doesn't do the work.
You can test them using the "docs", first example.

Commit:
changes to "coincidence" routine

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -0,0 +1,369 @@
+def rep(k, p):
+    lamda = k
+    pho = p[lamda]
+    while p != lamda:
```
Should this be `pho != lamda` (or `rho`?)

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -0,0 +1,369 @@
+def coincidence(C, alpha, beta):
+    l = 0
+    q = []
+    #XXX what are p, q here?
```
`p` could be an argument, or it could also be a part of `C`.

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -0,0 +1,369 @@
+    while i <= l:
+        gamma = q[i]
+        i += 1
+        del C[gamma]
```
It seems that this is already implicitly done by `merge` when setting `p[v] = mu`.

Commit:
Add example for "scan", code fix for Coincidence routine

Add the complete step of "Coincidence" working and "l" assignment

Commit:
Add the example 5.4 from Derek Holt

Commit:
Added the example for "scan" method from original todd-coxeter paper

gxyd commented:
I have added one of the examples mentioned in the originial todd-cox paper. In that paper, there doesn't happen any use of `coincidence` routine (`coincidence` routine is never reached). I am still seeing to improve over the `coincidence` routine.

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -0,0 +1,668 @@
+     [None, 2, None, 0, 2, 0, None, 2, 1, None],
+     [0, None, 0, None, None, None, None, 0, 0, 0]]
+
+    # Example from original Todd-Cox
```
Preferably Todd-Coxeter

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -0,0 +1,668 @@
+    C[alpha][A.index(x)] = beta
+    C[beta][A.index(x.inverse())] = alpha
+
+p = []
```
I think this is logically a part of the table together with the action data. We should probably have `class CosetTable`.

gxyd replied:
Ah! I was skeptical in making it a class, to be true. Now you have 
raised the point. I think I will try to code it first.

On 05/28/2016 11:49 AM, Kalevi Suominen wrote:
>
> In sympy/combinatorics/fp_groups.py 
> <https://github.com/sympy/sympy/pull/11140#discussion_r64984707>:
>
> > +#   their inverses), represented using "FpGroupElm"
> > +
> > +# Pg. 153
> > +## later: remove "A" from argument
> > +def define(C, alpha, x, A):
> > +    # now C.n is not obtainable
> > +    if len(C) == DefaultMaxLimit:
> > +        # abort the further generation of cosets
> > +        return
> > +    C.append([None]*len(A))
> > +    # beta is the new coset generated
> > +    beta = len(C) - 1
> > +    C[alpha][A.index(x)] = beta
> > +    C[beta][A.index(x.inverse())] = alpha
> > +
> > +p = []
>
> I think this is logically a part of the table together with the action 
> data. We should probably have |class CosetTable|.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/sympy/sympy/pull/11140/files/5f6f6b3f66ceef5bd48b68bbcb2ebb61855b64f6#r64984707>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/AHWt8Y6YwLsTBNvNU18_dFJu5hD_7sD8ks5qF959gaJpZM4IjQRZ>.
>

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -0,0 +1,668 @@
+    q = []
+    l = merge(alpha, beta, p, q, l)
+    i = 0
+    while i < l:
```
Perhaps we could think of  `q`  as a queue (as the choice of letter seems to hint). This could then be `while len(q) > 0:`, and the next line could be `gamma = q.pop(0)`. `i` and `l` would not be needed.

Commit:
Make CosetTable a class, add examples for HLT strategy

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -69,503 +71,514 @@ def coset_table(self, Y):
+        self.fp_group = fp_grp
+        self.subgroup = subgroup
+        self.p = [0]
+        self._A = list(chain.from_iterable((gen, gen**-1) \
```
Perhaps this could be `self.A`.

Commit:
CosetTable is "not" a subclass of `list`

Commit:
loop `scan_and_fill` until it has filled the alpha row in the table

also add examples for the HLT enumeration method

Commit:
Add cyclic_conjugates method for FreeGroupElements

Commit:
use "q" as a queue in "coincidence" routine

This improves the efficiency (is logically clear):
```
In [1]: F, a, b = free_group("a, b")
In [2]: Cox = FpGroup(F, [a**6, b**6, (a*b)**2, (a**2*b**2)**2, (a**3*b**3)**5])

"without" using "q" as queue
In [3]: %timeit f()
Out [3]: 1 loop, best of 3: 6.85 s per loop

"with" using "q" as queue
In [4]: %timeit f()
1 loop, best of 3: 6.6 s per loop
```

Commit:
use "A_dict" instead of "A.index" for finding index of (gen, gen^-1)

Commit:
Add test file for "Coset Enumeration"

gxyd commented:
I have added the test file. Though haven't added extensive tests. Will resolve the `.subword(i, i+1)` issue we discussed.
Will seek next improve using the "HLT" method given in the "1973chwd" paper.

Commit:
changes in Coset Table based method

Commit:
Add "Switch" method,also add "define_f"(for Felsch specific requirements)

Commit:
Make compress() routine work fine

gxyd commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -0,0 +1,553 @@
+                    self.table[gamma][A.index(x)] = beta
+                    self.table[beta][A.index(x**-1)] == gamma
+        # all the cosets in the table are live cosets
+        self.p = list(range(gamma + 1))
```
Does this assignment look dangerous? (its change of `id` of `p`)

jksuom replied:
I don't think so. `p` is not a part of the external interface.

Commit:
fix "standardize" method of Coset Table (for some still raises index error)

Commit:
FreeGroupElm -> FreeGroupElement (Aaron's comment fix)

Commit:
Add code for `coset_enumeration_c`

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -0,0 +1,595 @@
+        while i < r and self.table[f][A_dict[word.subword(i, i+1)]] is not None:
+            f = self.table[f][A_dict[word.subword(i, i+1)]]
+            i += 1
+        # can this be replaced with i == r ?
```
I believe it could be.

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -0,0 +1,595 @@
+            # run the "coincidence" routine
+            # line: We assume that before each call of COINCIDENCE, for α ∈ [1..n],
+            # we have p[α]=α iff α ∈ Ω.
+            self.p.extend(list(range(len(self))))
```
This line seems odd to me.

gxyd replied:
Yes, I know this line is to be removed. The other method `scan_and_fill` doesn't have this line.

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -0,0 +1,595 @@
+                for w in R:
+                    self.scan(beta, w)
+                    if p[beta] < beta:
+                        continue
```
What does this do? Should this be `break`?

Commit:
"stadardization" of a Coset Table

Commit:
remove wrong-code line from "scan" routine

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -0,0 +1,707 @@
+            if i >= r:
+                if f != alpha:
+                    self.coincidence(f, alpha)
+                    return
```
This `return` should probably be indented the same as `if f ...`. So we are finished in any case.

Commit:
Make indexing of word to return a generator or its inverse

```
>>> F, x, y = free_group("x, y")
>>> w = x**2*y**-4*x**5
>>> w[1]
x
>>> w[4]
y**-1
```
Also re-indent the "return" statement (according to comment) by @jksuom
[https://github.com/sympy/sympy/pull/11140/files/f67951c394b7102a6899696e95c913f811f72aec#r65673875]

Commit:
In `scan_and_fill` set first defition of "b" and "j" outisde loop

This makes `scan_and_fill` efficient, since it apparently always scans
to the end of a word even if part of the end has already been scanned
backwards.
So avoid re-scanning of the part of word

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -0,0 +1,715 @@
+        j = r - 1
+        while i_A < l_A:
+            # do the forward scanning
+            while i < r and self.table[f][A_dict[word[i]]] is not None:
```
This could probably be `while i <= j and ...` to end the loop if the indices meet.

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -0,0 +1,715 @@
+            while i < r and self.table[f][A_dict[word[i]]] is not None:
+                f = self.table[f][A_dict[word[i]]]
+                i += 1
+            if i >= r:
```
And here `i > j`.

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -0,0 +1,715 @@
+                f = self.table[f][A_dict[word[i]]]
+                i += 1
+            if i >= r:
+                if f != alpha:
```
Here we would then have `b` instead of the fixed `alpha`.

Commit:
change the comparison of cosets in `scan_and_fill`

Commit:
simplification to a tuple of "chi"

Commit:
Address Aaron's comment from PR#11350, Felsch improvements

Commit:
changes in methods for Felsch strategy, add doctest for Felsch strategy

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -0,0 +1,724 @@
+    @property
+    def n(self):
+        """The number `n` represents the largest number that has been used so
+        far for a live coset
```
It seems that `n` is actually not used for a live coset because of `+1`. Maybe this should be something like the length of the sublist containing the live cosets.

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -0,0 +1,724 @@
+    # Pg. 153
+    def define(self, alpha, x):
+        A = self.A
+        if self.n == CosetTableDefaultMaxLimit:
```
Maybe `len(self.table)` should be taken instead of `self.n` or else this should be an inequality. It is conceivable that the table could grow bigger than the limit with the tail containing only non-live cosets, and then a new live `beta` would be added making `self.n > CosetTableDefaultMaxLimit`.

gxyd replied:
I agree with making this `len(self.table)`.

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -0,0 +1,724 @@
+        i_A = 0
+        b = alpha
+        j = r - 1
+        while i_A < l_A:
```
This could probably be `while True:`. The loop is expected to run until `i > j`.

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -0,0 +1,724 @@
+            for w in R:
+                self.scan(beta, w)
+                if p[beta] < beta:
+                    continue
```
This could probably be `break`.

Commit:
documentation and code fixes related to coset enumeration

Commit:
fix the bug in compress() routine

Now verified every coset of table with GAP, seems to be working
fine.

Commit:
Add modifications, tests for Felsch strategy

Add tests for Feslch strategy and improvements

gxyd commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -503,15 +503,12 @@ def compress(self):
                 # replace α by γ in coset table
                 for x in A:
                     beta = self.table[alpha][A_dict[x]]
-                    if beta == alpha:
```
This line introduced the bug, its not required as the below this does the task. I have now verified "Mennickie group" (n=1, n=2), Cox group, "BH Neuman", "Burnside groups" (some more groups) from GAP for exact coset value (not just `index`) and these seem to be working good.

gxyd commented:
Now, apart from a few printing issues. I don't think much major changes are necessary in this PR. Perhaps the "Feslch" strategy tests are too slow.

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -0,0 +1,716 @@
+    @property
+    def n(self):
+        """The number `n` represents length of the sublist containing the live
+        cosets far for a live coset.
```
Maybe "The number 'n' represents the length of the sublist containing the live cosets".

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -0,0 +1,716 @@
+            # comment by Kalevi, this is already done by p[v] = mu
+            # del C[gamma]
+            # commenting this out
+            # omega = omega - set(gamma)
```
I think these comments could be dropped now. It seems that the deletion was some kind of typo in the original pseudocode.

Commit:
delete un-necessary comments from fp_groups.py

jksuom commented on sympy/combinatorics/fp_groups.py:
```diff
@@ -0,0 +1,684 @@
+                if p[alpha] < alpha:
+                    break
+            if p[alpha] < alpha:
+                continue
```
If `p[alpha] < alpha`, the while loop is repeated and only then `alpha += 1` is executed. Changing `continue` to `pass` and adding `else:` before the `for` loop would increment `alpha` straight away.

Commit:
changes in printing and combined the tests for two strategies

Commit:
Make FpGroup, CosetTable non-Basic just like FpGroup
